File: src/main/java/org/hisp/dhis/rules/models/RuleVariableNewestEvent.java
package org.hisp.dhis.rules.models;

import com.google.auto.value.AutoValue;
import org.hisp.dhis.rules.Option;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.RuleVariableValueMapBuilder;
import org.hisp.dhis.rules.Utils;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.hisp.dhis.rules.Utils.getLastUpdateDate;
import static org.hisp.dhis.rules.Utils.getLastUpdateDateForHistory;

@AutoValue
public abstract class RuleVariableNewestEvent
    extends RuleVariableDataElement
{

    @Nonnull
    public static RuleVariableNewestEvent create(@Nonnull String name,
                                                 @Nonnull String dataElement, @Nonnull RuleValueType dataElementValueType, boolean useCodeForOptionSet, List<Option> options)
    {
        return new AutoValue_RuleVariableNewestEvent( name, useCodeForOptionSet, options, dataElement, dataElementValueType );
    }

    @Override
    public Map<String, RuleVariableValue> createValues( RuleVariableValueMapBuilder builder,
        Map<String, List<RuleDataValueHistory>> allEventValues,
        Map<String, RuleAttributeValue> currentEnrollmentValues,
        Map<String, RuleDataValue> currentEventValues )
    {
        Map<String, RuleVariableValue> valueMap = new HashMap();
        List<RuleDataValueHistory> ruleDataValues = allEventValues.get( this.dataElement() );

        if ( ruleDataValues == null || ruleDataValues.isEmpty() )
        {
            valueMap.put( this.name(), RuleVariableValue.create( this.dataElementType() ) );
        }
        else
        {
            RuleVariableValue variableValue;

            RuleDataValueHistory value = ruleDataValues.get( 0 );

            String optionValue = this.useCodeForOptionSet() ? value.getValue() : getOptionName( value.getValue() );

            variableValue = RuleVariableValue.create( optionValue,
                this.dataElementType(), Utils.valuesForHistory( ruleDataValues ), getLastUpdateDateForHistory( ruleDataValues ) );


            valueMap.put( this.name(), variableValue );
        }

        return valueMap;
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleVariableNewestEvent.kt
package org.hisp.dhis.rules.models

import org.hisp.dhis.rules.*

class RuleVariableNewestEvent(
    val name: String,
    val useCodeForOptionSet: Boolean,
    val options: List<Option>,
    val  dataElement: String,
    val dataElementType: RuleValueType
) : RuleVariableDataElement {
    override fun dataElement(): String {
        return dataElement
    }

    override fun dataElementType(): RuleValueType {
        return dataElementType
    }

    override fun options(): List<Option> {
        return options
    }

    override fun createValues(
        builder: RuleVariableValueMapBuilder,
        allEventValues: Map<String, List<RuleDataValue>>,
        currentEnrollmentValues: Map<String, RuleAttributeValue>,
        currentEventValues: Map<String, RuleDataValue>
    ): Map<String, RuleVariableValue> {
        val valueMap: MutableMap<String, RuleVariableValue> = HashMap()
        val ruleDataValues = allEventValues[dataElement]
        if (ruleDataValues == null || ruleDataValues.isEmpty()) {
            valueMap[name] = RuleVariableValue(dataElementType)
        } else {
            val variableValue: RuleVariableValue
            val value = ruleDataValues[0]
            val optionValue = if (useCodeForOptionSet) value.value else getOptionName(value.value)!!
            variableValue = RuleVariableValue(
                dataElementType, optionValue,
                values(ruleDataValues), getLastUpdateDate(ruleDataValues)
            )
            valueMap[name] = variableValue
        }
        return valueMap
    }
}
